go/service/simplefs.go

```diff
@@ -12,6 +14,67 @@ import (
 	"golang.org/x/net/context"
 )
 
+// transporterWaiter exposes a get() method to get a transporter, and waits for
+// it if it's not available yet. Additionally, it also tries to have get()
+// calls from different goroutines return in the same order they come in. This
+// is done by assigning a number to each call, and process them one-by-one.
+// Only the current processing goroutine actively checks and waits for the
+// transporter. Other calls won't start until their previous ones have
+// finished.
+type transporterWaiter struct {
+	getter func() rpc.Transporter
+
+	// Just use atomic instead of having cond protect this, to minimize the
+	// chance that multiple calls wait on the same Mutex (which is not
+	// necessarily FIFO).
+	lastWaitingNumber uint64
+
+	cond                *sync.Cond
+	lastProcessedNumber uint64
+}
+
+func newTransporterWaiter(
+	tranporterGetter func() (transporter rpc.Transporter),
+) *transporterWaiter {
+	return &transporterWaiter{
+		getter:              tranporterGetter,
+		lastWaitingNumber:   0,
+		cond:                sync.NewCond(&sync.Mutex{}),
+		lastProcessedNumber: 0,
+	}
+}
+
+const waitForTransporterInterval = time.Second / 4
+const waitForTransporterTimeout = time.Minute
+
+func (w *transporterWaiter) get(ctx context.Context) (xp rpc.Transporter) {
+	myNumber := atomic.AddUint64(&w.lastWaitingNumber, 1)
+
+	ctx, cancel := context.WithTimeout(ctx, waitForTransporterTimeout)
+	defer cancel()
+
+	w.cond.L.Lock()
+	defer w.cond.L.Unlock()
+	for w.lastProcessedNumber != myNumber-1 {
+		w.cond.Wait()
+	}
+
+	xp = w.getter()
+loop:
+	for xp == nil {
+		time.Sleep(waitForTransporterInterval)
+		xp = w.getter()
+		select {
+		case <-ctx.Done():
+			break loop
```

Comment:
It is indeed because KBFS is not running though, by the time we fail it. If we just return the context error, it'd be hard to tell the difference between a timeout here and a timeout in the KBFS process. Perhaps `KBFSNotRunningError` should just be renamed, and wrap the ctx error so it indicates whether it's a cancel or timeout.

Reply:
Why is it a sure thing that KBFS is not running?  That doesn't seem to follow.  It could be that KBFS started running during the sleep, but the `select` case sees that the context is done and returns from there first.  Then you'd be getting the wrong error back -- the client would technically be expecting the context canceled error in that case.

I think if the context has been canceled or timed out, it's best to get the context error back.  I don't see any reason not to return it here.  If you really want to distinguish where the timeout happened, you could wrap the error with `github.com/pkg/errors` to include the full stack.